server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.FilterWeight;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.common.lucene.search.function.MinScoreScorer;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.format;
+
+public class VectorSimilarityQuery extends Query {
+    private final float similarity;
+    private final float docScore;
+    private final Query innerKnnQuery;
+
+    public VectorSimilarityQuery(Query innerKnnQuery, float similarity, float docScore) {
```

Comment:
super minor: would it make sense to limit the type of query that is provided as an argument to `AbstractKnnVectorQuery`, given that that is what we always provide to it?

Reply:
An, `AbstractKnnVectorQuery` is package private.

Reply:
good point :)